File: Windows/AndroidMic/Library/Audio/AudioBuffer.cs
using System;
using System.Threading;
using System.Threading.Tasks;

namespace AndroidMic.Audio
{
    // implementation of circular buffer
    public class AudioBuffer
    {
        public const int Capacity = 5 * 1024;
        public readonly byte[] Buffer = new byte[Capacity];

        private int regionSize = 0;
        private int regionLeft = 0;
        private int regionRight = 0;

        // can be accessed from at most one thread at a time
        private readonly SemaphoreSlim semaphore = new SemaphoreSlim(1, 1);

        ~AudioBuffer()
        {
            semaphore.Dispose();
        }

        /// <summary>
        /// Open buffer region to read from
        /// </summary>
        /// <param name="requestSize"></param>
        /// <param name="actualSize"></param>
        /// <param name="offset"></param>
        public async Task<Tuple<int, int>> OpenReadRegion(int requestSize)
        {
            await semaphore.WaitAsync();
            int actualSize = Math.Min(Math.Min(requestSize, regionSize), Capacity - regionLeft);
            int offset = regionLeft;
            return new Tuple<int, int>(actualSize, offset);
        }

        /// <summary>
        /// Close buffer read region
        /// </summary>
        /// <param name="readSize"></param>
        public void CloseReadRegion(int readSize)
        {
            regionLeft = (regionLeft + readSize) % Capacity;
            regionSize -= Math.Min(readSize, regionSize);
            semaphore.Release();
        }

        /// <summary>
        /// Open buffer region to write to
        /// </summary>
        /// <param name="requestSize"></param>
        /// <param name="actualSize"></param>
        /// <param name="offset"></param>
        public async Task<Tuple<int, int>> OpenWriteRegion(int requestSize)
        {
            await semaphore.WaitAsync();
            int actualSize = Math.Min(Math.Min(requestSize, Capacity - regionSize), Capacity - regionRight);
            int offset = regionRight;
            return new Tuple<int, int>(actualSize, offset);
        }

        /// <summary>
        /// Close buffer write region
        /// </summary>
        /// <param name="writeSize"></param>
        public void CloseWriteRegion(int writeSize)
        {
            regionRight = (regionRight + writeSize) % Capacity;
            regionSize = Math.Min(regionSize + writeSize, Capacity);
            semaphore.Release();
        }

        /// <summary>
        /// Get buffer size
        /// </summary>
        /// <returns></returns>
        public int Size()
        {
            semaphore.Wait();
            var result = regionSize;
            semaphore.Release();
            return result;
        }

        /// <summary>
        /// Whether buffer is empty
        /// </summary>
        /// <returns></returns>
        public bool IsEmpty()
        {
            semaphore.Wait();
            var result = regionSize == 0;
            semaphore.Release();
            return result;
        }

        /// <summary>
        /// Clear buffer
        /// </summary>
        public void Clear()
        {
            semaphore.Wait();
            regionLeft = 0;
            regionRight = 0;
            regionSize = 0;
            semaphore.Release();
        }
    }
}

File: Android/app/src/main/java/com/example/androidMic/domain/audio/AudioBuffer.kt
package com.example.androidMic.domain.audio

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlin.math.min

// thread safe circular buffer
class AudioBuffer {
    private var regionLeft = 0
    private var regionRight = 0
    private var regionSize = 0

    val capacity = 5 * 1024
    val buffer = ByteArray(capacity)

    private val mutex = Mutex()

    // get memory region to read from
    suspend fun openReadRegion(requestSize: Int): Pair<Int, Int> {
        mutex.lock()
        val actualSize = min(min(requestSize, regionSize), capacity - regionLeft)
        val offset = regionLeft
        return Pair(actualSize, offset)
    }

    // mark this region as read
    fun closeReadRegion(actualSize: Int) {
        regionLeft = (regionLeft + actualSize) % capacity
        regionSize -= min(actualSize, regionSize)
        mutex.unlock()
    }

    // get memory region to write to
    suspend fun openWriteRegion(requestSize: Int): Pair<Int, Int> {
        mutex.lock()
        val actualSize = min(min(requestSize, capacity - regionSize), capacity - regionRight)
        val offset = regionRight
        return Pair(actualSize, offset)
    }

    // mark this region as written
    fun closeWriteRegion(actualSize: Int) {
        regionRight = (regionRight + actualSize) % capacity
        regionSize = min(regionSize + actualSize, capacity)
        mutex.unlock()
    }

    // get valid buffer size
    suspend fun size(): Int {
        mutex.withLock {
            return regionSize
        }
    }

    // check if buffer is empty
    suspend fun isEmpty(): Boolean {
        mutex.withLock {
            return regionSize == 0
        }
    }

    suspend fun clear() {
        mutex.withLock {
            regionLeft = 0
            regionRight = 0
            regionSize = 0
        }
    }
}